tests/unit/Entity/EntityIdValueTest.php

```diff
@@ -75,41 +78,95 @@ public function testGetSortKey( EntityIdValue $id ) {
 		$this->assertInternalType( 'string', $id->getSortKey() );
 	}
 
+	public function provideGetArrayValue() {
+		return [
+			'Q2147483647' => [
+				new ItemId( 'Q2147483647' ),
+				[
+					'entity-type' => 'item',
+					'numeric-id' => 2147483647,
+					'id' => 'Q2147483647'
+				],
+			],
+			'P31337' => [
+				new PropertyId( 'P31337' ),
+				[
+					'entity-type' => 'property',
+					'numeric-id' => 31337,
+					'id' => 'P31337',
+				],
+			],
+			'X567' => [
+				$this->newCustomId( 'X567' ),
+				[
+					'entity-type' => 'custom',
+					'id' => 'X567',
+				],
+			],
+			'foo:P678' => [
+				new PropertyId( 'foo:P678' ),
+				[
+					'entity-type' => 'property',
+					'numeric-id' => 678,
+					'id' => 'foo:P678',
+				],
+			],
+		];
+	}
+
 	/**
-	 * @dataProvider instanceProvider
+	 * @dataProvider provideGetArrayValue
 	 */
-	public function testGetArrayValueRoundtrip( EntityIdValue $id ) {
-		$newId = EntityIdValue::newFromArray( $id->getArrayValue() );
+	public function testGetArrayValue( EntityId $id, array $expected ) {
+		$value = new EntityIdValue( $id );
+		$array = $value->getArrayValue();
 
-		$this->assertEquals( $id, $newId );
+		ksort( $expected );
+		ksort( $array );
+		$this->assertEquals( $expected, $array );
```

Comment:
You are right regarding PHP vs. abstract data model. 
I disagree with your conclusion though.
The interface should not give guarantees no needed by the caller, based on the current implementation. 
Guaranteeing this order makes calling code more brittle, and the code itself harder to maintain.

Point in case is this patch. This test fails without the ksort. So should I make the code more complex to uphold a pointless guarantee, which is not stated anywhere, and which no production code relies on (I hope)? Or should I just drop the assumption, and make the test ablivious to the order? The latter seems easier, and gives more flexibility for the future.

Interface guarantees should never include anything that just happens to hold for the current implementation. An interface's contract needs to strike a balance between the stability needed by the callers, and the flexibility wanted to change the implementation. Needlessly restricting that flexibility is not a good thing.